core_types/src/signature.rs

```diff
@@ -2,17 +2,28 @@
 //! a way of providing cryptographically verifiable proof of a given agent
 //! as having been the author of a given data entry.
 
-use crate::cas::content::Address;
+use crate::{cas::content::Address, error::error::HolochainError, json::JsonString};
 
-/// Provenance is a tuple of initiating agent and signature of some item being signed
+/// Provenance is a tuple of initiating agent public key and signature of some item being signed
 /// this type is used in headers and in capability requests where the item being signed
 /// is implicitly known by context
-pub type Provenance = (Address, Signature);
-/*impl Provenance {
-    fn source(&self) -> Address {
-        self.1.clone()
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Hash, Eq, DefaultJson)]
+pub struct Provenance {
```

Comment:
this could add "a lot" of redundant data if we're serialising the same `source` and `signature` keys and sending them around

use the new ergonomics methods and leave the tuple in place?

Reply:
good catch, fixed.